### PR TITLE
feat: scheduler plugins (5/): add cluster eligibility plugin logic

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR is part of the PRs that implement the in-tree scheduler framework plugins.

It features the cluster eligibility plugin.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests

### CODE REVIEW

**Comment by michaelawyu** [comment]
The eligibility check is extracted into its own package as another controller uses it (the member cluster scheduler source controller); which will be sent in a separate PR.

File: pkg/scheduler/clustereligibilitychecker/checker.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+// Package clustereligibilitychecker features a utility for verifying if a member cluster is
+// eligible for resource placement.
+package clustereligibilitychecker
```
Comment (zhiying-lin) [inline pkg/scheduler/clustereligibilitychecker/checker.go:8]:
```suggestion
package clustereligibilityvalidator
```

niiit, how about using validator?

File: pkg/scheduler/clustereligibilitychecker/checker.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+	// defaultClusterHealthCheckTimeout is the default timeout value this checker uses for checking
+	// if a cluster is still in a healthy state.
+	defaultClusterHealthCheckTimeout = time.Minute * 5
+)
+
+type ClusterEligibilityChecker struct {
+	// The timeout value this checker uses for checking if a cluster has been disconnected
+	// from the fleet for a prolonged period of time.
+	clusterHeartbeatTimeout time.Duration
```
Comment (zhiying-lin) [inline pkg/scheduler/clustereligibilitychecker/checker.go:32]:
```suggestion
	clusterHeartbeatCheckTimeout time.Duration
```

niiiit, make it consistent with "healthCheckTimeout"?

File: pkg/scheduler/clustereligibilitychecker/checker.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		// a) the member cluster object spec is most of the time not touched after creation; and
+		// b) as long as the heartbeat signal does not timeout, a little drift in genrations
+		//    should not exclude a cluster from resource scheduling.
+		return false, "cluster is not connected to the fleet: member agent not joined yet"
+	}
+
+	memberAgentHealthyCond := cluster.GetAgentCondition(fleetv1beta1.MemberAgent, fleetv1beta1.AgentHealthy)
+	if memberAgentHealthyCond == nil {
+		// The health condition is absent.
+		return false, "cluster is not connected to the fleet: health condition not available"
```
Comment (zhiying-lin) [inline pkg/scheduler/clustereligibilitychecker/checker.go:129]:
```suggestion
		return false, "cluster is not connected to the fleet: member agent health status is in the unknown"
```

to indicate it's member agent health status.

File: pkg/scheduler/framework/plugins/clustereligibility/plugin.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	defaultPluginName = "ClusterEligibility"
+)
+
+// Plugin is the scheduler plugin that performs the cluster eligibility check.
+type Plugin struct {
+	// The name of the plugin.
+	name string
+
+	// The cluster eligibility checker in use by the scheduler framework.
+	clusterEligibilityChecker *clustereligibilitychecker.ClusterEligibilityChecker
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/plugins/clustereligibility/plugin.go:29]: nit, hmm, is it necessary?

use the handle.ClusterEligibilityChecker() directly?

File: pkg/scheduler/clustereligibilitychecker/checker.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/scheduler/clustereligibilitychecker/checker.go:8]: I have some reservations; `validator` assumes that we are verifying the validity of the eligibility, which is not the true purpose of this package.

File: pkg/scheduler/framework/plugins/clustereligibility/plugin.go (hunk 4 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/plugins/clustereligibility/plugin.go]: Done.

File: pkg/scheduler/clustereligibilitychecker/checker.go (hunk 2 above)
Comment (michaelawyu) [reply pkg/scheduler/clustereligibilitychecker/checker.go]: Done.

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**